gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskConfiguration.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.compute;
+
+import com.google.api.services.compute.model.Disk;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.MoreObjects.ToStringHelper;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * Base class for Google Compute Engine disk configurations. A disk can be used as primary storage
+ * for your virtual machine instances. Use {@link StandardDiskConfiguration} to create a standard
+ * disk given a disk type and size. Use {@link ImageDiskConfiguration} to create a disk from a
+ * Compute Engine disk image. Use {@link SnapshotDiskConfiguration} to create a disk from a Compute
+ * Engine disk snapshot.
+ *
+ * @see <a href="https://cloud.google.com/compute/docs/disks/">Block Storage</a>
+ */
+public abstract class DiskConfiguration implements Serializable {
+
+  private static final long serialVersionUID = -1783061701255428417L;
+
+  private final Type type;
+  private final Long sizeGb;
+  private final DiskTypeId diskType;
+
+  /**
+   * Type of a Google Compute Engine disk configuration.
+   */
+  public enum Type {
+    /**
+     * A Google Compute Engine standard disk configuration.
+     */
+    STANDARD,
+
+    /**
+     * A Google Compute Engine disk configuration that creates a disk from an image.
+     */
+    IMAGE,
+
+    /**
+     * A Google Compute Engine disk configuration that creates a disk from a snapshot.
+     */
+    SNAPSHOT
+  }
+
+  /**
+   * Base builder for disk configurations.
+   *
+   * @param <T> the disk configuration type
+   * @param <B> the disk configuration builder
+   */
+  public abstract static class Builder<T extends DiskConfiguration, B extends Builder<T, B>> {
+
+    private Type type;
+    private Long sizeGb;
+    private DiskTypeId diskType;
+
+    Builder(Type type) {
+      this.type = type;
+    }
+
+    Builder(DiskConfiguration diskConfiguration) {
+      this.type = diskConfiguration.type;
+      this.sizeGb = diskConfiguration.sizeGb;
+      this.diskType = diskConfiguration.diskType;
+    }
+
+    Builder(Type type, Disk diskPb) {
+      this.type = type;
+      this.sizeGb = diskPb.getSizeGb();
```

Comment:
Snapshots and images have their size depending on the disk/raw source from which they have been taken. If this size is not specified the image/snapshot is used. If this is specified and is bigger than the size of the image/snapshot the the image/snapshot [will be resized](https://cloud.google.com/compute/docs/disks/persistent-disks#restoresnapshotlargersize).
If this value is set and is smaller than the image/snapshot size then disk creation will fail.

I'll update the docs to include this information.